lib/mlir/Dialect/Clift/Utils/ImportModel.cpp

```diff
@@ -0,0 +1,589 @@
+#include "llvm/Support/FormatVariadic.h"
+
+#include "revng/ADT/RecursiveCoroutine.h"
+
+#include "revng-c/mlir/Dialect/Clift/IR/Clift.h"
+#include "revng-c/mlir/Dialect/Clift/IR/CliftAttributes.h"
+#include "revng-c/mlir/Dialect/Clift/IR/CliftTypes.h"
+#include "revng-c/mlir/Dialect/Clift/Utils/ImportModel.h"
+
+namespace {
+
+namespace clift = mlir::clift;
+
+template<typename Attribute>
+using AttributeVector = llvm::SmallVector<Attribute, 16>;
+
+class CliftConverter {
+  mlir::MLIRContext *Context;
+  llvm::function_ref<mlir::InFlightDiagnostic()> EmitError;
+
+  llvm::DenseMap<uint64_t, clift::TypeDefinition> Cache;
+  llvm::DenseMap<uint64_t, const model::Type *> IncompleteTypes;
+
+  llvm::SmallSet<uint64_t, 16> DefinitionGuardSet;
+
+  class RecursiveDefinitionGuard {
+    CliftConverter *Self = nullptr;
+    uint64_t ID;
+
+  public:
+    explicit RecursiveDefinitionGuard(CliftConverter &Self, const uint64_t ID) {
+      if (Self.DefinitionGuardSet.insert(ID).second) {
+        this->Self = &Self;
+        this->ID = ID;
+      }
+    }
+
+    RecursiveDefinitionGuard(const RecursiveDefinitionGuard &) = delete;
+    RecursiveDefinitionGuard &
+    operator=(const RecursiveDefinitionGuard &) = delete;
+
+    ~RecursiveDefinitionGuard() {
+      if (Self != nullptr) {
+        Self->DefinitionGuardSet.erase(ID);
+      }
+    }
+
+    explicit operator bool() const { return Self != nullptr; }
+  };
+
+public:
+  explicit CliftConverter(mlir::MLIRContext &Context,
+                          llvm::function_ref<mlir::InFlightDiagnostic()>
+                            EmitError) :
+    Context(&Context), EmitError(EmitError) {}
+
+  CliftConverter(const CliftConverter &) = delete;
+  CliftConverter &operator=(const CliftConverter &) = delete;
+
+  ~CliftConverter() {
+    revng_assert(DefinitionGuardSet.empty());
+    revng_assert(IncompleteTypes.empty());
+  }
+
+  clift::ValueType convertUnqualifiedType(const model::Type &ModelType) {
+    const clift::ValueType T = getUnwrappedValueType(ModelType,
+                                                     /*RequireComplete=*/true);
+    if (T and not processIncompleteTypes()) {
+      IncompleteTypes.clear();
+      return nullptr;
+    }
+    return T;
+  }
+
+  clift::ValueType convertQualifiedType(const model::QualifiedType &ModelType) {
+    const clift::ValueType T = getQualifiedValueType(ModelType,
+                                                     /*RequireComplete=*/true);
+    if (T and not processIncompleteTypes()) {
+      IncompleteTypes.clear();
+      return nullptr;
+    }
+    return T;
+  }
+
+private:
+  mlir::BoolAttr getBool(bool const Value) {
+    return mlir::BoolAttr::get(Context, Value);
+  }
+
+  mlir::BoolAttr getFalse() { return getBool(false); }
+
+  template<typename T, typename... ArgTypes>
+  T make(const ArgTypes &...Args) {
+    if (failed(T::verify(EmitError, Args...)))
+      return {};
+    return T::get(Context, Args...);
+  }
```

Comment:
Isn't calling this equivalent to just calling `T::getChecked(Context, Args...)`? If so this is basically a just a helper for conciseness right?

Reply:
Yes. It used to be a little different, but I got rid of that and now it might be equivalent to just calling `getChecked`, except that `getChecked` here doesn't compile for all types. I didn't investigate it deeply enough to be sure on the root cause of that issue.